langchain/src/chat_models/baiduwenxin.ts

```diff
@@ -0,0 +1,425 @@
+import { BaseChatModel, BaseChatModelParams } from "./base.js";
+import {
+  AIChatMessage,
+  BaseChatMessage,
+  ChatGeneration,
+  ChatResult,
+  MessageType,
+} from "../schema/index.js";
+import fetchAdapter from "../util/axios-fetch-adapter.js";
+import type { StreamingAxiosConfiguration } from "../util/axios-types.js";
+import { CallbackManagerForLLMRun } from "../callbacks/manager.js";
+import { BaseLanguageModelCallOptions } from "../base_language/index.js";
+import { getEnvironmentVariable, isNode } from "../util/env.js";
+
+const axios = (
+  "default" in axiosMod ? axiosMod.default : axiosMod
+) as AxiosStatic;
+
+export type WenxinMessageRole = "assistant" | "user";
+
+interface WenxinMessage {
+  role: WenxinMessageRole;
+  content: string;
+}
+
+interface TokenUsage {
+  completionTokens?: number;
+  promptTokens?: number;
+  totalTokens?: number;
+}
+
+interface ChatCompletionRequest {
+  messages: WenxinMessage[];
+  stream?: boolean;
+  user_id?: string;
+  temperature?: number;
+  top_p?: number;
+  penalty_score?: number;
+}
+
+interface ChatCompletionResponse {
+  id: string;
+  object: string;
+  created: number;
+  result: string;
+  need_clear_history: boolean;
+  usage: TokenUsage;
+}
+
+declare interface BaiduWenxinChatInput {
+  /** Model name to use
+   * @default "ERNIE-Bot-turbo"
+   */
+  modelName: string;
+
+  /** Whether to stream the results or not. Defaults to false. */
+  streaming?: boolean;
+
+  /** Messages to pass as a prefix to the prompt */
+  prefixMessages?: WenxinMessage[];
+
+  /**
+   * ID of the end-user who made requests.
+   */
+  userId?: string;
+
+  /**
+   * API key to use when making requests. Defaults to the value of
+   * `BAIDU_ERNIE_API_KEY` environment variable.
+   */
+  baiduErnieApiKey?: string;
+
+  /**
+   * Secret key to use when making requests. Defaults to the value of
+   * `BAIDU_ERNIE_SECRET_KEY` environment variable.
+   */
+  baiduErnieSecretKey?: string;
+
+  /** Amount of randomness injected into the response. Ranges
+   * from 0 to 1 (0 is not included). Use temp closer to 0 for analytical /
+   * multiple choice, and temp closer to 1 for creative
+   * and generative tasks. Defaults to 0.95.
+   * Only supported for `modelName` of `WenxinModelName.ERNIE_BOT`.
+   */
+  temperature?: number;
+
+  /** Total probability mass of tokens to consider at each step. Range
+   * from 0 to 1.0. Defaults to 0.8.
+   * Only supported for `modelName` of `WenxinModelName.ERNIE_BOT`.
+   */
+  topP?: number;
+
+  /** Penalizes repeated tokens according to frequency. Range
+   * from 1.0 to 2.0. Defaults to 1.0.
+   * Only supported for `modelName` of `WenxinModelName.ERNIE_BOT`.
+   */
+  penaltyScore?: number;
+}
+
+function messageTypeToWenxinRole(type: MessageType): WenxinMessageRole {
+  switch (type) {
+    case "ai":
+      return "assistant";
+    case "human":
+      return "user";
+    case "system":
+      throw new Error("System messages not supported");
```

Comment:
Should these maybe just convert to `human` messages? A lot of chains/examples use system messages

Reply:
I'm afraid that it can not be converted to `human` directly. The official doc of ERNIE says: 

> 聊天上下文信息……必须为奇数个成员，成员中message的role必须依次为user、assistant   (https://cloud.baidu.com/doc/WENXINWORKSHOP/s/4lilb2lpf)

It means:

> Chat context information...Must be an odd number of messages, and the role of messages must be user, assistant in sequence

If throwing an Exception is bad, I have three other solutions:

1. Ignore the `system` message. Delete it from the array. It maybe mislead the caller that it is accepted.
2. Merge the `system` message to the latest `human` message. E.g. :
    3. System message: "You are a programmer"
    4. Latest human message: "Write a hello world in Python"
    5. Merged human message: "You are a programmer\n\nWrite a hello world in Python"
3. Convert it to `human` message and make an empty `AI` message. E.g. :
    3. System message: "You are a programmer". 
    4. Latest human message: "Write a hello world in Python".
    5. The last messages to ERNIE:
        1. Human message: "You are a programmer". 
        2. AI message: "".
        3. Human message: "Write a hello world in Python".

Which one is acceptable? Do you have other solutions? Are there any other LLMs who don't support system message?

Reply:
Ah dang. Let's just leave it then.

Google vertex has similar restrictions but allows the first message to be a system message only.